File: 0099-Recover Binary Search Tree/0099-Recover Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode* root) {
        if (root == nullptr)
        {
            return;
        }
        
        TreeNode* first = nullptr;
        TreeNode* second = nullptr;
        TreeNode* curr = root;
        TreeNode* prev = nullptr;
        stack<TreeNode*> St;
        while (!St.empty() || curr)
        {
            if (curr)
            {
                St.push(curr);
                curr = curr->left;
            }
            else
            {
                curr = St.top();
                St.pop();
                if (prev && curr->val < prev->val)
                {
                    if (!first)
                    {
                        first = prev;
                    }
                    second = curr;
                }
                prev = curr;
                curr = curr->right;
            }
        }
        
        swap(first->val, second->val);
    }
};

File: 0001-0100/0099-Recover Binary Search Tree/0099-Recover Binary Search Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        St = []
        first = second = prev = None
        curr = root
        while St or curr:
            if curr:
                St.append(curr)
                curr = curr.left
            else:
                curr = St.pop()
                if prev and prev.val > curr.val:
                    if first is None:
                        first = prev
                    second = curr
                prev = curr
                curr = curr.right
        first.val, second.val = second.val, first.val
